kart/s3_util.py

```diff
@@ -12,37 +14,82 @@
 
 # Utility functions for dealing with S3 - not yet launched.
 
+# Lots of functions in this file look like this:
+# >> @functools.lru_cache()
+# >> def _get_some_thread_unsafe_s3_resource(region, thread_hash):
+# This is a simple way of effectively getting a thread-local cache, ie,
+# - we want to get a resource that is configured how we currently need it (has the correct default region)
+# - we don't want a resource that is also being used by another thread
+# - aside from these two constraints, we can reuse the resources, hence the lru_cache.
 
-@functools.lru_cache(maxsize=1)
-def get_s3_config():
-    # TODO - add an option --s3-region to commands where it would be useful.
-    return None
+L = logging.getLogger("kart.s3_util")
 
 
-@functools.lru_cache(maxsize=1)
-def get_s3_client():
-    client = boto3.client("s3", config=get_s3_config())
+@functools.lru_cache()
+def get_region(bucket):
+    # The region -> bucket cache is not thread-local, since the region a bucket is in doesn't depend
+    # on which thread we are currently using.
+    if not bucket:
+        return None
+    try:
+        response = get_s3_client().head_bucket(Bucket=bucket)
+        return response["ResponseMetadata"]["HTTPHeaders"]["x-amz-bucket-region"]
+    except Exception as e:
```

Comment:
The main situation this might fail in seems to be if we have `GetObject` but not `ListBucket` permissions - in which case we can fetch the tiles but not get the bucket details. Is that right?

Do we require the `ListBucket` perm elsewhere for read-only access anyway, or is it only required for importing tiles? If we already need it elsewhere we might as well drop the except clause here

Reply:
We would require it if the user tried to use wildcards but we wouldn't require it if they listed in full the object(s) that make up the dataset. I'm inclined to leave this in since it just makes Kart a bit more robust